Upgrade smart contract tutorial

DOBEN opened the pull request:
## Purpose

closes #https://github.com/Concordium/concordium-rust-smart-contracts/issues/260

- Add upgrade smart contract tutorial

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/index.rst:
```diff
@@ -0,0 +1,28 @@
+Upgrading smart contracts on Concordium
+=======================================
+
+In this tutorial, you are going to get familiar with how to upgrade a smart contract natively. You explore how to correctly
```
```suggestion
In this tutorial, you are going to get familiar with how to upgrade a smart contract natively. You explore how to
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/index.rst:
```diff
@@ -0,0 +1,28 @@
+=======================================
+
+In this tutorial, you are going to get familiar with how to upgrade a smart contract natively. You explore how to correctly
+migrate the smart contract state from your old contract to the upgraded contract.
```
```suggestion
migrate the smart contract state correctly from your old contract to the upgraded contract.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/index.rst:
```diff
@@ -0,0 +1,28 @@
+In this tutorial, you are going to get familiar with how to upgrade a smart contract natively. You explore how to correctly
+migrate the smart contract state from your old contract to the upgraded contract.
+
+An important desired features of blockchains is that smart contracts once deployed are immutable by default.
```
```suggestion
An important desired features of blockchains is that smart contracts, once deployed, are immutable by default.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/index.rst:
```diff
@@ -0,0 +1,28 @@
+The past showed that development teams are looking for mutable smart contract options as well if they want to have
+the capabilities to fix bugs or to have the option to add additional
+features to their smart contract protocol in the future.
+Other blockchain require complex ``proxy-implementation`` patterns to achieve upgradability on the by default immutable smart contracts.
```
```suggestion
Other blockchains require complex ``proxy-implementation`` patterns to achieve upgradability on the, by default, immutable smart contracts.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+Native upgradability
+====================
+
+The goal of this tutorial is to deploy a ``contract-version1``, upgrade its logic and migrate its state to ``contract-version2``.
```
```suggestion
The goal of this tutorial is to deploy a ``contract-version1``, upgrade its logic, and migrate its state to ``contract-version2``.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+
+The `contract-version1 <https://github.com/Concordium/concordium-rust-smart-contracts/tree/main/examples/smart-contract-upgrade/contract-version1>`_
+includes an ``upgrade`` function and the `contract-version2 <https://github.com/Concordium/concordium-rust-smart-contracts/tree/main/examples/smart-contract-upgrade/contract-version2>`_
+includes a ``migration`` function which the next two sections explore.
```
```suggestion
includes a ``migration`` function; these functions are explored in the next two sections.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+``migration`` function
+======================
+
+A ``migration`` function can be optionally invoked as part of the upgrade to e.g. change the shape of the smart contract state.
```
```suggestion
A ``migration`` function can be optionally invoked as part of the upgrade to, e.g., change the shape of the smart contract state.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+======================
+
+A ``migration`` function can be optionally invoked as part of the upgrade to e.g. change the shape of the smart contract state.
+Use the below input file to specify the ``<EntryPointName>`` and the ``<InputParameters>``
```
```suggestion
Use the input file below to specify the ``<EntryPointName>`` and the ``<InputParameters>``
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+      "module": "<ModuleReferenceContractVersion2>"
+   }
+
+If the module reference should be upgraded without invoking a ``migration`` function, use the below input file instead:
```
```suggestion
If the module reference should be upgraded without invoking a ``migration`` function, use the input file below instead:
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+   If you want to ensure a smart contract is truly immutable you need to do your own research
+   and study its logic.
+
+The function is marked as ``low_level``. This is ``necessary`` since the high-level mutable functions store the state of the contract at the end of
```
```suggestion
The function is marked as ``low_level``. This is *necessary* since the high-level mutable functions store the state of the contract at the end of
```

Bargsteen replied:
I think the example should show the `low_level` attribute explicitly to make the context more clear here.

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+   and study its logic.
+
+The function is marked as ``low_level``. This is ``necessary`` since the high-level mutable functions store the state of the contract at the end of
+execution. This conflicts with migration since the shape of the state ``might`` be changed by the ``migration`` function. If the state is then written
```
```suggestion
execution. This conflicts with migration since the shape of the state *might* be changed by the ``migration`` function. If the state is then written
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+execution. This conflicts with migration since the shape of the state ``might`` be changed by the ``migration`` function. If the state is then written
+by this function it would overwrite the state stored by the ``migration`` function.
+
+Greater control is given to the smart contract developer when using ``low-level`` functions (e.g. additional state manipulation
```
```suggestion
Greater control is given to the smart contract developer when using ``low-level`` functions (e.g., additional state manipulation
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+
+.. note::
+
+   Using ``high-level`` function in your smart contract code (except for ``upgrade``/``migration`` functions)
```
```suggestion
   Using ``high-level`` functions in your smart contract code (except for ``upgrade``/``migration`` functions)
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+.. warning::
+
+   Since more responsibility is given to the smart contract developer, smart contract ``upgrade``/``migration`` functions
+   should be coded and executed by experts that know about the underlying risk. You are about to become such an expert.
```
```suggestion
   should be coded and executed by experts that know about the underlying risks. You are about to become such an expert.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+
+.. note::
+
+   You can remove the ``low-level`` attribute at the ``upgrade`` function, in other words, use a ``high-level`` function, in
+   the case that the shape of the state stays the same, in other words, no ``migration`` is executed. This means when the ``State`` struct in
+   ``contract-version1`` and ``contract-version2`` is identical, you don't need a ``migration`` function and you can remove the ``low-level`` attribute
```
```suggestion
   ``contract-version1`` and ``contract-version2`` is identical, you don't need a ``migration`` function, and you can remove the ``low-level`` attribute
```

Bargsteen replied:
I find this paragraph a bit difficult to read. Consider revising it. 
Perhaps just removing one of the `in other words`.

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+   Since more responsibility is given to the smart contract developer, smart contract ``upgrade``/``migration`` functions
+   should be coded and executed by experts that know about the underlying risk. You are about to become such an expert.
+
+You have the required knowledge now to execute the upgrade flow on testnet in the next section.
```
```suggestion
You now have the required knowledge to execute the upgrade flow on testnet in the next section.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+
+    $cd ./examples/smart-contract-upgrade/contract-version1
+
+Compile the smart contract to a wasm module which you will deploy to testnet in the next step.
```
```suggestion
Compile the smart contract to a Wasm module which you will deploy to testnet in the next step.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+   for the above placeholder ``YourAccountAddress`` can be found in :ref:`Setup the development environment<setup-env>`.
+
+Since other users will read this tutorial, you will likely get an error that the above module is already deployed.
+If the module is already deployed or not deployed does not make a
```
```suggestion
If the module is already deployed or not deployed it does not make a
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+
+   $concordium-client contract init <ModuleReference> --contract smart_contract_upgrade --energy 30000 --sender <YourAccountAddress> --grpc-port 10000  --grpc-ip node.testnet.concordium.com
+
+For example, when using the module reference from the above screenshot, the command will look as follows:
```
```suggestion
For example, when using the module reference from the image above, the command will look as follows:
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+   :alt: Initialize contract version 1
+   :align: center
+
+As seen in the screenshot above, this command returns your smart contract index ``4462``.
```
```suggestion
As seen in the image above, this command returns your smart contract index ``4462``.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+
+   $concordium-client contract invoke <YourContractVersion1Index> --entrypoint view --grpc-port 10000 --grpc-ip node.testnet.concordium.com
+
+For example, when using the smart contract index from the above screenshot, the command will look as follows:
```
```suggestion
For example, when using the smart contract index from the image above, the command will look as follows:
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+   :alt: View contract version 1
+   :align: center
+
+You need to compile and deploy ``contract-version2`` before we can upgrade our smart contract instance ``contract-version1``.
```
```suggestion
You need to compile and deploy ``contract-version2`` before you can upgrade your smart contract instance ``contract-version1``.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+
+    $cd ./examples/smart-contract-upgrade/contract-version2
+
+Compile the smart contract to a wasm module which you will deploy to testnet in the next step.
```
```suggestion
Compile the smart contract to a Wasm module which you will deploy to testnet in the next step.
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+      "module": "<ModuleReferenceContractVersion2>"
+   }
+
+For example, when using the module reference from the above screenshot, the file will look as follows:
```
```suggestion
For example, when using the module reference from the image above, the file will look as follows:
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+
+   $concordium-client contract update <YourContractVersion1Index> --entrypoint upgrade --parameter-json upgrade.json --energy 5000 --sender <YourAccountAddress> --grpc-port 10000 --grpc-ip node.testnet.concordium.com
+
+For example, when using the smart contract index from the above screenshot, the command will look as follows:
```
```suggestion
For example, when using the smart contract index from the image above, the command will look as follows:
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+   :alt: View contract version 2
+   :align: center
+
+If the migration function had a bug and changed the shape of the state wrongly, your state might be corrupted. You might see different output errors in case of a corrupted state. One example is shown below:
```
```suggestion
If the migration function had a bug and changed the shape of the state incorrectly, your state might be corrupted. You might see different output errors in case of a corrupted state. One example is shown below:
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/index.rst:
```diff
@@ -32,6 +32,8 @@ In the :ref:`sponsored Transactions <sponsoredTransactions>` tutorial, you will
 message in the |bw| (or |mw-gen2| that uses WalletConnect) and how to send the signature to a backend.
 You will then instruct the backend to submit the transaction on behalf of the user on-chain.
 
+In the :ref:`Smart Contract Upgrade <intro-smart-contract-upgrade>` tutorial you will deploy a smart contract and upgrade its logic. During the upgrade you will call a migration function to .
```
I think something got cut off at the end of this sentence.

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/index.rst:
```diff
@@ -0,0 +1,28 @@
+
+An important desired features of blockchains is that smart contracts once deployed are immutable by default.
+The past showed that development teams are looking for mutable smart contract options as well if they want to have
+the capabilities to fix bugs or to have the option to add additional
```
```suggestion
the capabilities to fix bugs, or to have the option to add additional
```

dg-concordium commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/index.rst:
```diff
@@ -0,0 +1,28 @@
+migrate the smart contract state from your old contract to the upgraded contract.
+
+An important desired features of blockchains is that smart contracts once deployed are immutable by default.
+The past showed that development teams are looking for mutable smart contract options as well if they want to have
```
```suggestion
History has shown that development teams are looking for mutable smart contract options as well if they want to have
```

dg-concordium reviewed (approved):
A few changes.

abizjak commented on source/mainnet/net/guides/dapp-examples.rst:
```diff
@@ -61,7 +61,7 @@ Select an example to see more information about it, such as a hosted dApp for yo
     `Demo front end <https://voting.testnet.concordium.com>`__ where you can try the functionality on Concordium's testnet
 
-    :ref: `Tutorial about the Voting dApp<voting-dapp>`
+    :ref:`Tutorial about the Voting dApp<voting-dapp>`
```
Is this intended?

DOBEN replied:
Yes, the link was broken because of the extra whitespace. You can see similar links in the same file which all have NO whitespace.

abizjak commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/index.rst:
```diff
@@ -0,0 +1,28 @@
+the capabilities to fix bugs or to have the option to add additional
+features to their smart contract protocol in the future.
+Other blockchain require complex ``proxy-implementation`` patterns to achieve upgradability on the by default immutable smart contracts.
+Concordium makes this process easier by exposing an upgrade mechanism to natively upgrade the smart contract.
```
I think it is important to explicitly state that there can still be immutable contracts, and the upgradability is an opt-in.

abizjak commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,287 @@
+
+.. note::
+
+   If you don't find a ``host.upgrade(module)`` snippet in a smart contract code, you can be sure that this contract
+   can not be natively upgraded. The smart contract can still be part of a ``proxy-implementation`` pattern.
```
So this is technically incorrect. They can directly invoke the primitive upgrade function. https://docs.rs/concordium-std/latest/concordium_std/prims/fn.upgrade.html

abizjak reviewed (approved):
Looks good.

Bargsteen reviewed (changes requested):
Nice work Doris!

I have a few minor suggestions and two larger ones.
1. I think the tutorial should explain how to _write_ a migration function, not just execute a migration function.
2. The `migration` and `upgrade` function sections seem a bit out of place as they are quite generic and therefore sound more like a guide. So I would move them to our upgrade guide.

But otherwise, nice work ;)

DOBEN commented:
@Bargsteen 
https://github.com/Concordium/concordium.github.io/pull/808#pullrequestreview-1413022103

**Addressed Point1** by adding the following explanation:

A ``migration`` function could have a variety of different purposes such as:

- Log some events.
- Change values in the state.
- Change the shape of the smart contract state.

   There is no universal ``migration`` function or you might decide to use no ``migration`` at all.
   In the example smart contract, a ``migration`` function is provided that should be suitable for most use cases. 
   In more detail, the example ``migration`` function 
   focuses on how to change the shape of your smart contract state since this bullet point from above is non-trivial.

**Addressed Point2** by adding the following explanation:
I added links between the `tutorial` and the general `upgrade guide`. Also added some explanation about the `low_level` attribute to the general `upgrade guide`. The general `upgrade guide` does not specify in detail how the ``migration`` function should look since there is no universal ``migration`` function that we want to promote. That is why I keep the `upgrade/migration` section in the tutorial as well but re-wrote them a bit.

Bargsteen commented on source/mainnet/smart-contracts/guides/upgradeable-contract.rst:
```diff
@@ -81,4 +92,8 @@ The JSON parameter for triggering an upgrade is of the form:
+
+.. seealso::
+
+   For a guide on how to send interact with a smart contract using JSON see :ref:`interact-instance-json-parameters`.
```
```suggestion
   For a guide on how to interact with a smart contract using JSON see :ref:`interact-instance-json-parameters`.
```

Bargsteen commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/index.rst:
```diff
@@ -0,0 +1,39 @@
+Upgrading smart contracts on Concordium
+=======================================
+
+In this tutorial, you are going to get familiar with how to upgrade a smart contract natively. You explore how to
```
```suggestion
In this tutorial, you are going to get familiar with how to upgrade a smart contract natively. You will explore how to
```

Bargsteen commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,330 @@
+==========================
+
+When deploying ``contract-version1``, you might not know the exact logic of the ``contract-version2`` yet.
+``Contract-versoin2`` might be a smart contract that fixes a security bug in your smart contract code which will be discovered at some point in the future.
```
```suggestion
``contract-version2`` might be a smart contract that fixes a security bug in your smart contract code which will be discovered at some point in the future.
```

DOBEN replied:
I left the capital letter since it is the beginning of the sentence but fixed the typo in version.

Bargsteen commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,330 @@
+
+When deploying ``contract-version1``, you might not know the exact logic of the ``contract-version2`` yet.
+``Contract-versoin2`` might be a smart contract that fixes a security bug in your smart contract code which will be discovered at some point in the future.
+For that reason, it is good to have the option to call a ``migration`` function in ``contract-versoin2`` when executing the upgrade mechanism in ``contract-version1``.
```
```suggestion
For that reason, it is good to have the option to call a ``migration`` function in ``contract-version2`` when executing the upgrade mechanism in ``contract-version1``.
```

Bargsteen commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,330 @@
+      ...
+   }
+
+The ``upgrade`` functions in the example smart contracts are marked as ``low_level``. This is *necessary* since ``high_level`` mutable functions store the state of the contract at the end of
```
Consider using a different formatting than `` as `high_level` isn't referring to something in the code.

Bargsteen reviewed (commented):
A few minor findings, but looks good 😊

DOBEN commented on source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst:
```diff
@@ -0,0 +1,330 @@
+   In more detail, the example ``migration`` function
+   focuses on how to change the shape of your smart contract state since this bullet point from above is non-trivial.
+
+``Contract-version1`` enables optionally to call a ``migration`` function as part of the upgrade.
```
I left the capital letter since it is the start of a sentence. Changed the rest.

Bargsteen replied:
It still refers to a specific name in the code which is all lower case, but alright. I can live with that ;)
